\section{genomic}
\label{module_genomic}
This module contains the following objects

\begin{itemize}
\item \ref{object_Genomic} Genomic

\item \ref{object_GenomicRepeat} GenomicRepeat

\item This module also contains some factory methods
\end{itemize}
\subsection{genomic factory methods}
\subsubsection{read_fasta_file_Genomic}
\begin{description}
\item[External C] {\tt Wise2_read_fasta_file_Genomic (filename,length_of_N)}
\item[Perl] {\tt &Wise2::read_fasta_file_Genomic (filename,length_of_N)}

\end{description}
Arguments
\begin{description}
\item[filename] [UNKN ] filename to be opened and read [char *]
\item[length_of_N] [UNKN ] length of N to be considered repeat. -1 means none [int]
\item[returns] [UNKN ] Undocumented return value [Genomic *]
\end{description}
Reads a fasta file assumming that it is Genomic. 
Will complain if it is not, and return NULL.


\subsubsection{read_fasta_Genomic}
\begin{description}
\item[External C] {\tt Wise2_read_fasta_Genomic (ifp,length_of_N)}
\item[Perl] {\tt &Wise2::read_fasta_Genomic (ifp,length_of_N)}

\end{description}
Arguments
\begin{description}
\item[ifp] [UNKN ] file point to be read from [FILE *]
\item[length_of_N] [UNKN ] length of N to be considered repeat. -1 means none [int]
\item[returns] [UNKN ] Undocumented return value [Genomic *]
\end{description}
Reads a fasta file assumming that it is Genomic. 
Will complain if it is not, and return NULL.


\subsubsection{Genomic_from_Sequence_Nheuristic}
\begin{description}
\item[External C] {\tt Wise2_Genomic_from_Sequence_Nheuristic (seq,length_of_N)}
\item[Perl] {\tt &Wise2::Genomic_from_Sequence_Nheuristic (seq,length_of_N)}

\end{description}
Arguments
\begin{description}
\item[seq] [UNKN ] Undocumented argument [Sequence *]
\item[length_of_N] [UNKN ] Undocumented argument [int]
\item[returns] [UNKN ] Undocumented return value [Genomic *]
\end{description}
makes a new genomic from a Sequence, but
assummes that all the N runs greater than
a certain level are actually repeats.


\subsubsection{Genomic_from_Sequence}
\begin{description}
\item[External C] {\tt Wise2_Genomic_from_Sequence (seq)}
\item[Perl] {\tt &Wise2::Genomic_from_Sequence (seq)}

\end{description}
Arguments
\begin{description}
\item[seq] [OWNER] Sequence to make genomic from [Sequence *]
\item[returns] [UNKN ] Undocumented return value [Genomic *]
\end{description}
makes a new genomic from a Sequence. It 
owns the Sequence memory, ie will attempt a /free_Sequence
on the structure when /free_Genomic is called


If you want to give this genomic this Sequence and
forget about it, then just hand it this sequence and set
seq to NULL (no need to free it). If you intend to use 
the sequence object elsewhere outside of the Genomic datastructure
then use Genomic_from_Sequence(/hard_link_Sequence(seq))


This is part of a strict typing system, and therefore
is going to convert all non ATGCNs to Ns. You will lose
information here.




\subsection{Object Genomic}

\label{object_Genomic}

The Genomic object has the following fields. To see how to access them refer to \ref{accessing_fields}
\begin{description}
\item{baseseq} Type [Sequence * : Scalar] No documentation

\item{repeat} Type [GenomicRepeat ** : List] No documentation

\end{description}
No documentation for Genomic

Member functions of Genomic

\subsubsection{truncate_Genomic}

\begin{description}
\item[External C] {\tt Wise2_truncate_Genomic (gen,start,stop)}
\item[Perl] {\tt &Wise2::Genomic::truncate_Genomic (gen,start,stop)}

\item[Perl-OOP call] {\tt $obj->truncate_Genomic(start,stop)}

\end{description}
Arguments
\begin{description}
\item[gen] [READ ] Genomic that is truncated [Genomic *]
\item[start] [UNKN ] Undocumented argument [int]
\item[stop] [UNKN ] Undocumented argument [int]
\item[returns] [UNKN ] Undocumented return value [Genomic *]
\end{description}
Truncates a Genomic sequence. Basically uses
the /magic_trunc_Sequence function (of course!)


It does not alter gen, rather it returns a new
sequence with that truncation


Handles repeat information correctly.


\subsubsection{Genomic_name}

\begin{description}
\item[External C] {\tt Wise2_Genomic_name (gen)}
\item[Perl] {\tt &Wise2::Genomic::Genomic_name (gen)}

\item[Perl-OOP call] {\tt $obj->Genomic_name()}

\end{description}
Arguments
\begin{description}
\item[gen] [UNKN ] Undocumented argument [Genomic *]
\item[returns] [UNKN ] Undocumented return value [char *]
\end{description}
Returns the name of the Genomic


\subsubsection{Genomic_length}

\begin{description}
\item[External C] {\tt Wise2_Genomic_length (gen)}
\item[Perl] {\tt &Wise2::Genomic::Genomic_length (gen)}

\item[Perl-OOP call] {\tt $obj->Genomic_length()}

\end{description}
Arguments
\begin{description}
\item[gen] [UNKN ] Undocumented argument [Genomic *]
\item[returns] [UNKN ] Undocumented return value [int]
\end{description}
Returns the length of the Genomic


\subsubsection{Genomic_seqchar}

\begin{description}
\item[External C] {\tt Wise2_Genomic_seqchar (gen,pos)}
\item[Perl] {\tt &Wise2::Genomic::Genomic_seqchar (gen,pos)}

\item[Perl-OOP call] {\tt $obj->Genomic_seqchar(pos)}

\end{description}
Arguments
\begin{description}
\item[gen] [UNKN ] Genomic [Genomic *]
\item[pos] [UNKN ] position in Genomic to get char [int]
\item[returns] [UNKN ] Undocumented return value [char]
\end{description}
Returns sequence character at this position.


\subsubsection{show_Genomic}

\begin{description}
\item[External C] {\tt Wise2_show_Genomic (gen,ofp)}
\item[Perl] {\tt &Wise2::Genomic::show (gen,ofp)}

\item[Perl-OOP call] {\tt $obj->show(ofp)}

\end{description}
Arguments
\begin{description}
\item[gen] [UNKN ] Undocumented argument [Genomic *]
\item[ofp] [UNKN ] Undocumented argument [FILE *]
\item[returns] Nothing - no return value
\end{description}
For debugging


\subsection{Object GenomicRepeat}

\label{object_GenomicRepeat}

The GenomicRepeat object has the following fields. To see how to access them refer to \ref{accessing_fields}
\begin{description}
\item{start} Type [int : Scalar] No documentation

\item{end} Type [int : Scalar] No documentation

\item{type} Type [char * : Scalar] No documentation

\end{description}
No documentation for GenomicRepeat

Member functions of GenomicRepeat

