\section{cdparser}
\label{module_cdparser}
This module contains the following objects

\begin{itemize}
\item \ref{object_cDNAParser} cDNAParser

\item This module also contains some factory methods
\end{itemize}
\subsection{cdparser factory methods}
\subsubsection{flat_cDNAParser}
\begin{description}
\item[External C] {\tt Wise2_flat_cDNAParser (p)}
\item[Perl] {\tt &Wise2::flat_cDNAParser (p)}

\end{description}
Arguments
\begin{description}
\item[p] [READ ] probability of an indel [Probability]
\item[returns] [UNKN ] Undocumented return value [cDNAParser *]
\end{description}
Makes a flat (ie, indels of 1 or 2 == p)
cDNA parser. This means that insertions
and deletions of both 1 or 2 bases are
all parameterised at the same probability






\subsection{Object cDNAParser}

\label{object_cDNAParser}

The cDNAParser object has the following fields. To see how to access them refer to \ref{accessing_fields}
\begin{description}
\item{trans[PCD_PARSER_TRANS_LEN]} Type [Probability : Scalar] No documentation

\end{description}
This object holds the (very few) extra
transition information needed for the
estwise algorithm. It is sort of like
the 'gene model' part of sequencing
error (but very very simple)






Member functions of cDNAParser

